### PR TITLE
Remove pointer math (comparison) from render caching, as it is unreliable

### PR DESCRIPTION
**Pull request by paulbellamy**
As mentioned in https://github.com/weaveworks/scope/issues/896#issuecomment-184173816

Might fix the flakey test? Not sure. Either way `reflect.Value.Pointer` of a func is unreliable.

The tradeoff is that we have individual caches for each renderer, and have to forward `ResetCache` calls throughout the tree, which is ugly.

### CODE REVIEW

**Comment by 2opremio** [comment]
The code/approach looks fine. However, I don't know enough about this part of the code to know whether it's complete or it's missing edge cases.

**Comment by paulbellamy** [comment]
I feel the same. Will have reviewed by @tomwilkie

**Comment by paulbellamy** [comment]
After in-person discussion fons suggests generating a stable/random ID for each renderer, and using a single-cache against that.

**Comment by 2opremio** [comment]
Much cleaner :+1:

That said, it won't work against mutable renderers. Are all renderers immutable?

**Comment by paulbellamy** [comment]
afaik, yes.

**Comment by tomwilkie** [comment]
> Are all renderers immutable?

They are all deterministic (which I think is the property we want here).

The renderers themselves are structs which are not inherently immutable, but we don't mutate them anywhere.

**Comment by tomwilkie** [comment]
> Might fix the flakey test?

I could get this to flake in a loop pretty easily.  Will test tomorrow with this fix.

File: render/memoise.go [hunk 1]
```diff
@@ -13,23 +13,22 @@ var renderCache = gcache.New(100).LRU().Build()
...
 func (m *memoise) Render(rpt report.Report) RenderableNodes {
-	key := ""
-	v := reflect.ValueOf(m.Renderer)
-	switch v.Kind() {
-	case reflect.Ptr, reflect.Func:
-		key = fmt.Sprintf("%s-%x", rpt.ID, v.Pointer())
-	default:
-		return m.Renderer.Render(rpt)
-	}
+	key := fmt.Sprintf("%s-%s", rpt.ID, m.id)
```
Comment (tomwilkie) [inline render/memoise.go]: +1 much nicer.

**Comment by tomwilkie** [comment]
Flakes reliably on master:
```
Toms-MacBook-Pro:app twilkie$ while ./app.test --test.run TestAll; do sleep 0.1; done
PASS
PASS
PASS
PASS
--- FAIL: TestAll (0.01s)
	scope_test.go:78: api_topology_test.go:39: Expected status 200, got 404. Path: /api/topology/containers-by-hostname/container%3A5e4d3c2b1a
FAIL
Toms-MacBook-Pro:app twilkie$ while ./app.test --test.run TestAll; do sleep 0.1; done
PASS
PASS
PASS
PASS
--- FAIL: TestAll (0.02s)
	scope_test.go:78: api_topology_test.go:39: Expected status 200, got 404. Path: /api/topology/containers-by-image/container%3A5e4d3c2b1a
FAIL
Toms-MacBook-Pro:app twilkie$ while ./app.test --test.run TestAll; do sleep 0.1; done
PASS
PASS
PASS
PASS
PASS
PASS
PASS
PASS
PASS
PASS
PASS
PASS
PASS
PASS
PASS
PASS
--- FAIL: TestAll (0.02s)
	scope_test.go:78: api_topology_test.go:39: Expected status 200, got 404. Path: /api/topology/containers-by-image/container_image%3Aimageid123
FAIL
```

Ran for 5mins on this branch with no flakes.

**Pull request merged**